### ISSUE TITLE
DSL for model definition

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#115]
#### Problem Statement

Currently, neural network construction in the library requires verbose, imperative code that lacks the expressiveness and readability needed for modern deep learning workflows. Users face several challenges when defining neural networks:

- **Verbose Network Construction**: Manual instantiation and chaining of modules creates lengthy, hard-to-read code
- **Error-Prone Architecture Definition**: Easy to make mistakes in layer connections and dimension matching
- **Limited Composability**: Difficult to create reusable network components and complex architectures
- **Poor Developer Experience**: Lacks the intuitive, declarative syntax that ML practitioners expect from modern frameworks
- **Type Safety Issues**: Manual type management across layers increases risk of runtime errors

#### Solution

Implement a comprehensive Kotlin DSL for neural network definition that provides an intuitive, declarative approach to building neural architectures. The DSL should offer:

1. **Declarative Layer Definition**: Clean, readable syntax for defining network layers
2. **Automatic Dimension Inference**: Smart layer chaining with automatic input/output dimension matching
3. **Type Safety**: Full Kotlin type safety with generic tensor types
4. **Composability**: Support for creating reusable network components and complex architectures
5. **Flexible Architecture Support**: Handle sequential networks, branched architectures, and custom topologies
6. **Built-in Layer Types**: Comprehensive support for common layer types (dense, convolutional, activation, etc.)

#### Implementation Components

The solution includes:

- `NeuralNetworkDsl<T, V>`: Main DSL interface for network definition
- `NetworkBuilder<T, V>`: Builder pattern implementation for network construction
- Layer-specific DSL interfaces: `DENSE<T, V>`, `FLATTEN<T, V>`, etc.
- Support classes: `DenseImpl`, `FlattenImpl`, `StageImpl` for internal implementation
- Type-safe generic support with convenient FP32 shortcuts

##### Basic Sequential Network

```kotlin
// Simple feedforward network
val simpleNetwork = network<FP32, Float> {
    input(inputSize = 784) // MNIST input size
    
    dense(outputDimension = 128) {
        // Custom weight initialization if needed
        weights { shape -> /* custom init */ }
        bias { shape -> /* custom bias init */ }
    }
    
    activation { it.relu() }
    
    dense(outputDimension = 64)
    activation { it.relu() }
    
    dense(outputDimension = 10) // Output layer
    activation { it.softmax(dimension = -1) }
}
```

##### Convenience FP32 Network

```kotlin
// Using FP32 convenience function
val networkFP32 = networkFP32 {
    input(784)
    
    dense(128)
    activation { it.relu() }
    
    dense(64)
    activation { it.relu() }
    
    flatten() // Flatten layer with default parameters
    
    dense(10)
    activation { it.softmax(dimension = -1) }
}
```

##### Complex Architecture with Stages

```kotlin
val complexNetwork = network<FP32, Float> {
    input(inputSize = 224 * 224 * 3) // RGB image input
    
    // Feature extraction stage
    stage("feature_extraction") {
        dense(512)
        activation { it.relu() }
        
        dense(256)
        activation { it.relu() }
    }
    
    // Classification head
    stage("classifier") {
        flatten()
        dense(128)
        activation { it.relu() }
        
        dense(10) // Number of classes
        activation { it.softmax(dimension = -1) }
    }
}
```

##### Sequential Blocks

```kotlin
val networkWithSequential = networkFP32 {
    input(784)
    
    // First block
    sequential {
        dense(256)
        activation { it.relu() }
        
        dense(128)
        activation { it.relu() }
    }
    
    // Second block
    sequential {
        dense(64)
        activation { it.relu() }
        
        dense(32)
        activation { it.relu() }
    }
    
    // Output layer
    dense(10)
    activation { it.softmax(dimension = -1) }
}
```

##### Custom Layer Configuration

```kotlin
val customNetwork = network<FP32, Float> {
    input(784)
    
    dense(outputDimension = 128, id = "hidden1") {
        // Xavier/Glorot initialization
        weights { shape -> 
            randomNormal(shape) * sqrt(2.0 / shape.dimensions[0])
        }
        // Zero bias initialization
        bias { shape -> zeros(shape) }
    }
    
    activation(id = "relu1") { it.relu() }
    
    dense(outputDimension = 10, id = "output")
    activation(id = "softmax") { it.softmax(dimension = -1) }
}
```

##### Advanced Network Patterns

```kotlin
// ResNet-like skip connections (conceptual)
val advancedNetwork = networkFP32 {
    input(512)
    
    stage("block1") {
        dense(512)
        activation { it.relu() }
        dense(512)
        // Skip connection would be handled by custom module
    }
    
    stage("block2") {
        dense(256)
        activation { it.relu() }
        dense(256)
    }
    
    flatten("global_flatten")
    dense(10)
    activation { it.softmax(dimension = -1) }
}
```

#### Benefits

- **Intuitive Syntax**: Declarative DSL that reads like natural language
- **Type Safety**: Full Kotlin type safety with generic tensor support
- **Readable Code**: Self-documenting network architecture definitions
- **Error Prevention**: Automatic dimension inference prevents common mistakes
- **Flexible**: Supports both simple sequential networks and complex architectures
- **Composable**: Reusable network components and stages
- **Performance**: Efficient internal representation with minimal overhead

#### Technical Requirements

- Kotlin Multiplatform support (Common, JVM, JS, Native)
- Generic support for different tensor data types (`DType` and value types)
- Integration with existing `Tensor<T, V>` and `Module<T, V>` interfaces
- Support for custom layer initialization functions
- Memory efficient network construction
- Thread-safe DSL operations

#### Network Architecture Support

- **Sequential Networks**: Linear chain of layers
- **Staged Networks**: Named groups of layers for organization
- **Dense (Fully Connected) Layers**: With configurable dimensions and initialization
- **Activation Functions**: ReLU, Sigmoid, Tanh, Softmax, and custom activations
- **Flatten Layers**: Configurable dimension flattening
- **Custom Layers**: Extensible architecture for user-defined layers

#### DSL Features

1. **Automatic Dimension Inference**: Smart input/output dimension matching between layers
2. **Named Components**: Optional IDs for layers and stages for debugging and visualization
3. **Custom Initialization**: User-defined weight and bias initialization functions
4. **Nested Composition**: Sequential blocks and stages for complex architectures
5. **Type Shortcuts**: Convenient `networkFP32` for common float32 networks
6. **Builder Pattern**: Internal builder pattern for efficient network construction

#### Acceptance Criteria

- [x] Implement `NeuralNetworkDsl<T, V>` interface with all layer types
- [x] Create `NetworkBuilder<T, V>` for efficient network construction
- [x] Support for `input`, `dense`, `flatten`, `activation` layers
- [x] Implement `sequential` and `stage` composition mechanisms
- [x] Add custom weight and bias initialization support
- [x] Create type-safe generic versions and FP32 convenience functions
- [x] Ensure automatic dimension inference between connected layers
- [x] Support for named layers and stages with optional IDs
- [x] Integration with existing `Module<T, V>` interface
- [x] Comprehensive test suite covering all DSL features
- [x] Documentation and usage examples for all supported patterns
- [x] Verify compatibility across all supported Kotlin platforms

#### Error Handling

- Compile-time type safety for tensor operations
- Runtime validation for dimension mismatches
- Clear error messages for invalid network configurations
- Graceful handling of initialization failures

#### Performance Considerations

- Efficient network construction with minimal object allocation
- Lazy evaluation where appropriate
- Memory-efficient internal representation
- Fast network instantiation for training and inference

This DSL will dramatically improve the developer experience for neural network construction, making the library more accessible to ML practitioners while maintaining the type safety and performance characteristics of the underlying tensor operations.

**Comment by michalharakal** [comment]
Closing as implemented. Model-definition DSL ships in `nn/dsl/NetworkBuilder.kt` (`network<>`, `sequential {}`, `stage()`, layer builders).
